File: ios/RecipeClipper/Data/Local/DefaultGroceryRepository.swift
import Combine
import Foundation

/// The real, SQLite-backed GroceryRepository (Android's DefaultGroceryRepository). Database
/// failures are logged; writes become no-ops, a delete returns nil and a read is empty.
final class DefaultGroceryRepository: GroceryRepository {
    private let db: AppDatabase
    private let clock: Clock
    private let decisions: DecisionRepository?

    init(db: AppDatabase, clock: Clock, decisions: DecisionRepository? = nil) {
        self.db = db
        self.clock = clock
        self.decisions = decisions
    }

    func observeItems() -> AnyPublisher<[GroceryItem], Never> {
        db.observe { conn in try GroceryDao(db: conn).items().map(\.domain) }
    }

    func add(_ lines: [NewGroceryLine]) async {
        let now = clock.now()
        let decided = await decisions?.current() ?? .none
        let items: [GroceryItemRecord] = lines.compactMap { line in
            let text = line.text.kTrimmed
            guard !text.isEmpty else { return nil }
            let words = LanguageWords.forTag(line.language)
            // The keyword table first; for what it puts in Other, an aisle the model decided (#104).
            var aisle = Aisles.of(text, words: words)
            if aisle == .other, let words, let name = IngredientName.of(text, words: words) {
                aisle = decided.aisle(name, language: words.language) ?? .other
            }
            return GroceryItemRecord(
                text: text, language: line.language,
                aisle: aisle.key,
                sortOrder: 0, // the DAO puts them last, in order
                recipeId: line.recipeId, plannedDay: line.plannedDay, updatedAt: now
            )
        }
        if items.isEmpty { return }
        await perform("addGroceries") { dao in try dao.add(items) }
    }

    func setChecked(_ ids: [Int64], checked: Bool) async {
        let now = clock.now()
        await perform("setGroceryChecked") { dao in try dao.setChecked(ids, checked: checked, now: now) }
    }

    func setAisle(_ ids: [Int64], aisle: Aisle) async {
        let now = clock.now()
        await perform("setGroceryAisle") { dao in try dao.setAisle(ids, aisle: aisle.key, now: now) }
    }

    func fileFromOther(_ ids: [Int64], aisle: Aisle) async {
        let now = clock.now()
        await perform("fileGroceryAisle") { dao in try dao.fileFromOther(ids, aisle: aisle.key, now: now) }
    }

    func delete(_ ids: [Int64]) async -> DeletedGroceries? {
        await remove("deleteGroceries") { dao in try dao.items(ids: ids) }
    }

    func clearChecked() async -> DeletedGroceries? {
        await remove("clearCheckedGroceries") { dao in try dao.checkedItems() }
    }

    private func remove(_ what: String, _ which: @escaping (GroceryDao) throws -> [GroceryItemRecord]) async -> DeletedGroceries? {
        do {
            return try await db.write { conn -> DeletedGroceries? in
                let dao = GroceryDao(db: conn)
                let items = try which(dao)
                if items.isEmpty { return nil }
                try dao.delete(items.map(\.id))
                return DeletedGroceries(items: items)
            }
        } catch {
            dataLog.error("\(what, privacy: .public) failed: \(String(describing: error), privacy: .public)")
            return nil
        }
    }

    func restore(_ deleted: DeletedGroceries) async {
        await perform("restoreGroceries") { dao in try dao.restore(deleted.items) }
    }

    func plannedIngredients(start: Int64, end: Int64) async -> [PlannedIngredients] {
        do {
            return try await db.read { conn in try GroceryDao(db: conn).plannedIngredients(start: start, end: end) }
        } catch {
            dataLog.error("plannedIngredients failed: \(String(describing: error), privacy: .public)")
            return []
        }
    }

    private func perform(_ what: String, _ body: @escaping (GroceryDao) throws -> Void) async {
        do {
            try await db.write { conn in try body(GroceryDao(db: conn)) }
        } catch {
            dataLog.error("\(what, privacy: .public) failed: \(String(describing: error), privacy: .public)")
        }
    }
}

extension GroceryItemRecord {
    var domain: GroceryItem {
        GroceryItem(
            id: id, text: text, language: language, aisle: Aisle.fromKey(aisle), checked: checked,
            sortOrder: sortOrder, recipeId: recipeId, plannedDay: plannedDay
        )
    }
}

File: app/src/main/java/com/example/recipeclipper/data/DefaultGroceryRepository.kt
package com.example.recipeclipper.data

import com.example.recipeclipper.data.local.dao.GroceryDao
import com.example.recipeclipper.data.local.dao.PlannedIngredientsRow
import com.example.recipeclipper.data.local.entity.GroceryItemEntity
import com.example.recipeclipper.data.model.Aisle
import com.example.recipeclipper.data.model.Aisles
import com.example.recipeclipper.data.model.GroceryItem
import com.example.recipeclipper.data.model.LanguageWords
import com.example.recipeclipper.data.model.NewGroceryLine
import com.example.recipeclipper.data.model.PlannedIngredients
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The real, Room-backed [GroceryRepository]. Like [DefaultListRepository], a database failure
 * is logged and degrades ([ErrorLog.guard]): writes become no-ops, a delete returns null and a
 * Flow emits an empty list.
 */
@Singleton
class DefaultGroceryRepository @Inject constructor(
    private val dao: GroceryDao,
    private val clock: Clock,
    private val log: ErrorLog
) : GroceryRepository {

    override fun observeItems(): Flow<List<GroceryItem>> =
        dao.observeItems().map { rows -> rows.map { it.toDomain() } }.orEmptyOnError(log, "observeGroceries")

    override suspend fun add(lines: List<NewGroceryLine>) {
        val now = clock.now()
        val items = lines.mapNotNull { line ->
            val text = line.text.trim().takeIf { it.isNotEmpty() } ?: return@mapNotNull null
            GroceryItemEntity(
                text = text,
                language = line.language,
                aisle = Aisles.of(text, LanguageWords.forTag(line.language)).key,
                sortOrder = 0, // the DAO puts them last, in order
                recipeId = line.recipeId,
                plannedDay = line.plannedDay,
                updatedAt = now
            )
        }
        if (items.isEmpty()) return
        log.guard("addGroceries", Unit) { dao.add(items) }
    }

    override suspend fun setChecked(ids: List<Long>, checked: Boolean) =
        log.guard("setGroceryChecked", Unit) { dao.setChecked(ids, checked, clock.now()) }

    override suspend fun setAisle(ids: List<Long>, aisle: Aisle) =
        log.guard("setGroceryAisle", Unit) { dao.setAisle(ids, aisle.key, clock.now()) }

    override suspend fun delete(ids: List<Long>): GroceryRepository.DeletedItems? = log.guard("deleteGroceries", null) {
        val entities = dao.items(ids).ifEmpty { return@guard null }
        dao.delete(entities.map { it.id })
        GroceryRepository.DeletedItems(entities)
    }

    override suspend fun clearChecked(): GroceryRepository.DeletedItems? = log.guard("clearCheckedGroceries", null) {
        val entities = dao.checkedItems().ifEmpty { return@guard null }
        dao.delete(entities.map { it.id })
        GroceryRepository.DeletedItems(entities)
    }

    override suspend fun restore(deleted: GroceryRepository.DeletedItems) =
        log.guard("restoreGroceries", Unit) { dao.restore(deleted.entities) }

    override suspend fun plannedIngredients(start: Long, end: Long): List<PlannedIngredients> =
        log.guard("plannedIngredients", emptyList()) { dao.plannedIngredients(start, end).map { it.toDomain() } }
}

private fun GroceryItemEntity.toDomain() = GroceryItem(
    id = id,
    text = text,
    language = language,
    aisle = Aisle.fromKey(aisle),
    checked = checked,
    sortOrder = sortOrder,
    recipeId = recipeId,
    plannedDay = plannedDay
)

private fun PlannedIngredientsRow.toDomain() = PlannedIngredients(
    entryId = entryId,
    day = day,
    servings = servings,
    recipeId = recipeId,
    title = title,
    ingredients = ingredients,
    yield = yield,
    language = language
)
